tarantool/crud.py

```diff
@@ -0,0 +1,81 @@
+"""
+This module provides API for interaction with a crud module.
+"""
+
+from tarantool.error import DatabaseError
+
+
+class CrudResponse(object):
+    """
+    Contains response fields from the `crud`_ module that correspond to the lua implementation.
+
+    .. _crud: https://github.com/tarantool/crud/
+    """
+
+    def __init__(self, response):
+        """
+        Sets response fields as in lua implementation.
+
+        :param response: The response object of the crud module call.
+        :type response: :class:`~tarantool.response.Response`
+        """
+
+        if isinstance(response, dict):
+            for response_field_name in response.keys():
+                if isinstance(response_field_name, bytes):
+                    setattr(self, response_field_name.decode(), response[response_field_name])
+                else:
+                    setattr(self, response_field_name, response[response_field_name])
+        elif isinstance(response, str) or isinstance(response, bytes):
+            self.str = response
+            self.err = response
+        else:
+            raise RuntimeError('Unable to decode response to object due to unknown type')
+
+
+class CrudResult(CrudResponse):
```

Comment:
I don't see any reason to implement `CrudResponse` class which serves two different purposes through if/else condition in init. Just put corresponding init parts to `CrudResult` and `CrudError`.

Reply:
I mean something like this

```
class CrudResult():
    def __init__(self, response):
        assert(response, dict)

        for response_field_name in response.keys():
            if isinstance(response_field_name, bytes):
                setattr(self, response_field_name.decode(), response[response_field_name])
            else:
                setattr(self, response_field_name, response[response_field_name])

class CrudError():
    def __init__(self, response):
        assert isinstance(response, (str, bytes))
        self.str = response
        self.err = response
```

since we never create `CrudResponse` objects: it is always either a `CrudResult` or a `CrudError`

Reply:
Unfortunately, the code above does not give the desired behavior. 
I can suggest such code for consideration, but it will duplicate some behavior in init:

```python
class CrudResult():
    """
    Contains result's fields from result variable 
    of crud module operation.
    """

    def __init__(self, response):
        """
        Sets response fields as in Lua implementation.

        :param response: The response object of the crud module call.
        :type response: :class:`~tarantool.response.Response`
        """

        if isinstance(response, dict):
            for response_field_name in response.keys():
                if isinstance(response_field_name, bytes):
                    setattr(self, response_field_name.decode(), response[response_field_name])
                else:
                    setattr(self, response_field_name, response[response_field_name])


class CrudError():
    """
    Contains error's fields from error variable 
    of crud module operation.
    """

    def __init__(self, response):
        """
        Sets response fields as in Lua implementation.

        :param response: The response object of the crud module call.
        :type response: :class:`~tarantool.response.Response`
        """

        if isinstance(response, dict):
            for response_field_name in response.keys():
                if isinstance(response_field_name, bytes):
                    setattr(self, response_field_name.decode(), response[response_field_name])
                else:
                    setattr(self, response_field_name, response[response_field_name])
        elif isinstance(response, str) or isinstance(response, bytes):
            self.str = response
            self.err = response
```

Reply:
Yep, it seems that the case was not needed:

```python
elif isinstance(response, str) or isinstance(response, bytes):
        self.str = response
        self.err = response
```

Then version of code proposed by me for consideration will look like this:

```python
class CrudResult():
    """
    Contains result's fields from result variable 
    of crud module operation.
    """

    def __init__(self, response):
        """
        Sets response fields as in Lua implementation.

        :param response: The response object of the crud module call.
        :type response: :class:`~tarantool.response.Response`
        """

        if isinstance(response, dict):
            for response_field_name in response.keys():
                if isinstance(response_field_name, bytes):
                    setattr(self, response_field_name.decode(), response[response_field_name])
                else:
                    setattr(self, response_field_name, response[response_field_name])


class CrudError():
    """
    Contains error's fields from error variable 
    of crud module operation.
    """

    def __init__(self, response):
        """
        Sets response fields as in Lua implementation.

        :param response: The response object of the crud module call.
        :type response: :class:`~tarantool.response.Response`
        """

        if isinstance(response, dict):
            for response_field_name in response.keys():
                if isinstance(response_field_name, bytes):
                    setattr(self, response_field_name.decode(), response[response_field_name])
                else:
                    setattr(self, response_field_name, response[response_field_name])
```

Reply:
Well, in this case it really makes sense to use a single class with inheritance.